### PR TITLE
Implement router tree Walker, an analogy to filepath.Walk

### PR DESCRIPTION
**Pull request by VojtechVitek**
Chi router tree walker:
```go
func Walk(r Routes, walkFn WalkFunc) error

type WalkFunc func(method string, route string, handler http.Handler, middlewares ...func(http.Handler) http.Handler) error
```

This is an analogy to [filepath.Walk](https://golang.org/pkg/path/filepath/#Walk) from stdlib.

---
Example:
```go
r := yourOwnRouter()

if err := chi.Walk(r, func(method string, route string, handler http.Handler, middlewares ...func(http.Handler) http.Handler) error {

	// Print all routes in format "GET /favicon.ico"
	fmt.Printf("%v %v\n", method, route)

	return nil
}); err != nil {
	// handle error coming from the above WalkFunc
}
```

You can use the Walker for automatic API docs generation, for printing out all the routes and their middleware stack or for testing purposes (ie. you can check whether specific set of routes have a ACL middleware applied etc.). Walk func lets you filter out certain methods/routes easily (ie. filter out non-public routes).

If this gets merged, I plan to build automatic RAML docgen in https://github.com/go-chi/docgen  based on this functionality.

### CODE REVIEW

File: walker.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
+package chi
```
Comment (pkieltyka) [inline walker.go]: please move this code to tree.go instead of a new file walker.go

**Comment by pkieltyka** [comment]
this looks great thank you! A few requests

- [ ] move walker code to tree.go
- [ ] remove older walk methods to remove code redundancy 
- [ ] lets merge this PR once a PR for go-chi/docgen is ready

**Comment by VojtechVitek** [comment]
TestWalker walks `bigMux()` router and prints method+route for each entry.

```
=== RUN   TestWalker
--- PASS: TestWalker (0.00s)
	tree_test.go:433: GET /
	tree_test.go:433: GET /favicon.ico
	tree_test.go:433: GET /folders/*/
	tree_test.go:433: GET /folders/*/public
	tree_test.go:433: GET /hubs/{hubID}/view
	tree_test.go:433: GET /hubs/{hubID}/view/*
	tree_test.go:433: POST /hubs/{hubSlug}/view/*
	tree_test.go:433: GET /hubs/*/{hubID}/*/
	tree_test.go:433: GET /hubs/*/{hubID}/*/posts/*/
	tree_test.go:433: GET /hubs/*/{hubID}/*/touch
	tree_test.go:433: PATCH /hubs/*/{hubID}/*/webhooks/*
	tree_test.go:433: POST /hubs/*/{hubID}/*/webhooks/*
	tree_test.go:433: HEAD /hubs/*/{hubID}/*/webhooks/*
	tree_test.go:433: GET /hubs/*/{hubID}/*/webhooks/*
	tree_test.go:433: LINK /hubs/*/{hubID}/*/webhooks/*
	tree_test.go:433: DELETE /hubs/*/{hubID}/*/webhooks/*
	tree_test.go:433: UNLINK /hubs/*/{hubID}/*/webhooks/*
	tree_test.go:433: TRACE /hubs/*/{hubID}/*/webhooks/*
	tree_test.go:433: PUT /hubs/*/{hubID}/*/webhooks/*
	tree_test.go:433: CONNECT /hubs/*/{hubID}/*/webhooks/*
	tree_test.go:433: OPTIONS /hubs/*/{hubID}/*/webhooks/*
	tree_test.go:433: GET /suggestions
	tree_test.go:433: GET /woot/{wootID}/*
PASS
ok  	github.com/go-chi/chi	0.011s
```

**Comment by hugows** [comment]
Hi @VojtechVitek,
Could you explain the meaning of the asterisk on the routes pattern? Like, this:

`go-chi\chi\_examples\router-walk>go run main.go
GET /
PUT /ping
GET /road/*/left
POST /road/*/right
`

seems to suggest that the router matches /road/**something**/left instead of /road/left?

**Comment by VojtechVitek** [comment]
The asterisk most likely represents a subrouter.

**Pull request merged**